src/Extension/AttributeExtension.php

```diff
@@ -0,0 +1,179 @@
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Twig\Extension;
+
+use Twig\Environment;
+use Twig\Extension\Attribute\AsTwigFilter;
+use Twig\Extension\Attribute\AsTwigFunction;
+use Twig\Extension\Attribute\AsTwigTest;
+use Twig\TwigFilter;
+use Twig\TwigFunction;
+use Twig\TwigTest;
+
+/**
+ * Define Twig filters, functions, and tests with PHP attributes.
+ *
+ * @author Jérôme Tamarelle <jerome@tamarelle.net>
+ */
+final class AttributeExtension extends AbstractExtension implements WithLastModified
+{
+    private array $filters;
+    private array $functions;
+    private array $tests;
+
+    public function __construct(
+        /**
+         * A list of objects or class names defining filters, functions, and tests using PHP attributes.
+         * When passing a class name, it must be available in runtimes.
+         *
+         * @var iterable<object|class-string>
+         */
+        private iterable $objectsOrClasses,
+    ) {
+    }
+
+    public function getLastModified(): int
+    {
+        $lastModified = 0;
+
+        foreach ($this->objectsOrClasses as $objectOrClass) {
+            $r = new \ReflectionClass($objectOrClass);
+            if (is_file($r->getFileName()) && ($extensionTime = filemtime($r->getFileName())) > $lastModified) {
+                $lastModified = $extensionTime;
+            }
+        }
+
+        return $lastModified;
+    }
+
+    public function getFilters(): array
+    {
+        if (!isset($this->filters)) {
+            $this->initFromAttributes();
+        }
+
+        return $this->filters;
+    }
+
+    public function getFunctions(): array
+    {
+        if (!isset($this->functions)) {
+            $this->initFromAttributes();
+        }
+
+        return $this->functions;
+    }
+
+    public function getTests(): array
+    {
+        if (!isset($this->tests)) {
+            $this->initFromAttributes();
+        }
+
+        return $this->tests;
+    }
+
+    private function initFromAttributes()
+    {
+        $filters = $functions = $tests = [];
+
+        foreach ($this->objectsOrClasses as $objectOrClass) {
+            try {
+                $reflectionClass = new \ReflectionClass($objectOrClass);
+            } catch (\ReflectionException $e) {
+                throw new \LogicException(sprintf('"%s" class requires a list of objects or class name, "%s" given.', __CLASS__, get_debug_type($objectOrClass)), 0, $e);
+            }
+
+            foreach ($reflectionClass->getMethods() as $method) {
+                // Filters
+                foreach ($method->getAttributes(AsTwigFilter::class) as $attribute) {
+                    $attribute = $attribute->newInstance();
+
+                    $name = $attribute->name;
+                    if (isset($filters[$name])) {
+                        throw new \LogicException(sprintf('Multiple definitions of the "%s" filter.', $name));
```

Comment:
Could we have a unique centralized system for that ? Because not all functions/filters will come from here i guess ?

Reply:
This is a feature of Twig. You can replace an existing filter/function/test by declaring the same name in an extension.
I'm thinking about removing this check.